File: csharp/Shrink.cs
// Trace shrinking via delta debugging (ddmin).
//
// Finding a seed that breaks an invariant is only half the job; a 40-fault schedule
// that triggers a bug is nearly useless to a human. ddmin repeatedly throws away
// subsets of the faults and keeps whichever smaller schedule still fails, converging
// on a minimal reproducer. This is the classic Zeller/Hildebrandt algorithm.

using System;
using System.Collections.Generic;

namespace SeedSim;

public static class Shrink
{
    private static List<List<T>> Split<T>(IReadOnlyList<T> xs, int n)
    {
        var k = xs.Count / n;
        var r = xs.Count % n;
        var outChunks = new List<List<T>>();
        var i = 0;
        for (var idx = 0; idx < n; idx++)
        {
            var size = k + (idx < r ? 1 : 0);
            if (size > 0)
            {
                var chunk = new List<T>();
                for (var j = 0; j < size; j++) chunk.Add(xs[i + j]);
                outChunks.Add(chunk);
            }
            i += size;
        }
        return outChunks;
    }

    /// <summary>
    /// Return a minimal sublist of <paramref name="items"/> for which
    /// <paramref name="stillFails"/> holds. <c>stillFails(items)</c> must be true to
    /// begin with - you only shrink a known failure.
    /// </summary>
    public static List<T> Ddmin<T>(IReadOnlyList<T> items, Func<List<T>, bool> stillFails)
    {
        var cur = new List<T>(items);
        if (!stillFails(cur))
            throw new ArgumentException("ddmin requires the full input to already fail");

        var n = 2;
        while (cur.Count >= 2)
        {
            var chunks = Split(cur, n);
            var reduced = false;
            for (var i = 0; i < chunks.Count; i++)
            {
                var complement = new List<T>();
                for (var j = 0; j < chunks.Count; j++)
                    if (j != i) complement.AddRange(chunks[j]);
                if (complement.Count > 0 && stillFails(complement))
                {
                    cur = complement;
                    n = Math.Max(n - 1, 2);
                    reduced = true;
                    break;
                }
            }
            if (!reduced)
            {
                if (n >= cur.Count) break;
                n = Math.Min(cur.Count, 2 * n);
            }
        }
        return cur;
    }
}

File: ts/src/shrink.ts
// Trace shrinking via delta debugging (ddmin).
//
// Finding a seed that breaks an invariant is only half the job; a 40-fault
// schedule that triggers a bug is nearly useless to a human. Ddmin repeatedly
// throws away subsets of the faults and keeps whichever smaller schedule still
// fails, converging on a minimal reproducer. This is the classic
// Zeller/Hildebrandt algorithm.

function split<T>(xs: T[], n: number): T[][] {
  const k = Math.floor(xs.length / n);
  const r = xs.length % n;
  const out: T[][] = [];
  let i = 0;
  for (let idx = 0; idx < n; idx++) {
    let size = k;
    if (idx < r) {
      size += 1;
    }
    if (size > 0) {
      out.push(xs.slice(i, i + size));
      i += size;
    }
  }
  return out;
}

/**
 * Returns a minimal sublist of `items` for which `stillFails` holds.
 *
 * `stillFails(items)` must be true to begin with - you only shrink a known
 * failure - otherwise this throws.
 */
export function ddmin<T>(items: T[], stillFails: (subset: T[]) => boolean): T[] {
  let cur = [...items];
  if (!stillFails(cur)) {
    throw new Error("ddmin requires the full input to already fail");
  }

  let n = 2;
  while (cur.length >= 2) {
    const chunks = split(cur, n);
    let reduced = false;
    for (let i = 0; i < chunks.length; i++) {
      const complement: T[] = [];
      for (let j = 0; j < chunks.length; j++) {
        if (j !== i) {
          complement.push(...chunks[j]);
        }
      }
      if (complement.length > 0 && stillFails(complement)) {
        cur = complement;
        n = n - 1 > 2 ? n - 1 : 2;
        reduced = true;
        break;
      }
    }
    if (!reduced) {
      if (n >= cur.length) {
        break;
      }
      n = 2 * n < cur.length ? 2 * n : cur.length;
    }
  }
  return cur;
}
